### ISSUE TITLE
Compilation problems

### ISSUE THREAD

**Issue by pgimeno** [Quent42340/OpenMiner#22]
I'm having issues when trying to compile OpenMiner.

- I needed to copy FindSFML.cmake from /usr/share/SFML/cmake/Modules to ./cmake/ for it to be found.
- The GameKit include files were not found while compiling, even though I specified `-DGAMEKIT_LIBRARIES` and `-DGAMEKIT_INCLUDE_DIR` and CMake found it. I worked around this by editing the top CMakeLists.txt and replacing `include_directories(${GAMEKIT_INCLUDE_DIRS})` (plural DIRS) with `include_directories(${GAMEKIT_INCLUDE_DIR})`. I don't know why this is necessary. I know nothing about cmake.
- After that I got the following error.
  > .../OpenMiner/common/source/core/Registry.cpp: In member function ‘Item& Registry::registerItem(u32, const string&, const string&)’:
  > .../OpenMiner/common/source/core/Registry.cpp:25:29: error: invalid initialization of non-const reference of type ‘Item&’ from an rvalue of type ‘void’
  return m_items.emplace_back(internalID, textureID, id, name);

Would you please advise me on how to get compilation to continue?

**Comment by Quent42340** [comment]
Hello @pgimeno,

It seems that you're using an old SFML version. Since 2.5 it no longer provides `FindSFML.cmake` but provides `/usr/lib/cmake/SFML/SFMLConfig.cmake` instead.

However you can open a PR to add `FindSFML.cmake` file to `./cmake/` folder if it helps to compile with older versions.

And yes, looks like I was using `DIRS` (which is usually the name in .cmake files) instead of `DIR`. This problem is fixed by 4bd5fc1.

For the last error, which compiler are you using? If you're using an old version of GCC, please use clang instead since it has better C++17 support. Nevertheless, I fixed this particular error in 2064a40, but you may encounter it somewhere else.

A small tip: before C++17 `emplace` and `emplace_back` were returning `void`, but since C++17 they return `T&` or `const T&`, which isn't supported by some compiler versions with experimental C++17 support. therefore I highly recommend GCC 8.x or clang 5.x.

**Comment by pgimeno** [comment]
Thank you. Maybe a C++17 compliant compiler should be listed as a dependency?

I'm using the toolchain in stock Debian stable. It ships with:

- g++ 6.3.0
- clang 3.8.1
- cmake 3.7.2
- libsfml 2.4.1

I guess I'll have to wait a few years before attempting to compile this. Thanks anyway.

**Comment by pgimeno** [comment]
Thanks for the recommendation in IRC of installing clang from squeeze-backports. I installed clang-6.0. The build progressed past these problems but then choked on  `#include <string_view>` which was not found.

**Comment by Quent42340** [comment]
Sorry for yesterday @pgimeno, I ended up not looking at my computer at all. :/

Can you describe more? Please send the full error from clang. I tried locally with clang

And I suggested you tried clang-5.0, because it's what I use in travis-ci and it's known to work, I don't know about clang-6.0 but you got an error so please try clang-5.0, it should work

Here's travis-ci config to compile on Ubuntu Trusty: https://github.com/Quent42340/OpenMiner/blob/master/.travis.yml

**Comment by pgimeno** [comment]
Finally got my hands on this, sorry for the delay. As discussed in IRC, this is due to using a stdc++ library with no C++17 support, despite the compiler being capable of compiling C++17.

I tried installing a buster chroot. The compilation ran all the way to the end, despite the many warnings. The server started successfully and printed `Hello from Lua!`, but the client won't start. Even `glxgears` doesn't run in the chroot, giving the same error.

```
$ LIBGL_DEBUG=verbose glxgears
libGL: screen 0 does not appear to be DRI2 capable
libGL: OpenDriver: trying /usr/lib/x86_64-linux-gnu/dri/tls/swrast_dri.so
libGL: OpenDriver: trying /usr/lib/x86_64-linux-gnu/dri/swrast_dri.so
libGL: Can't open configuration file /etc/drirc: No such file or directory.
libGL: Can't open configuration file $HOME/.drirc: No such file or directory.
libGL: Can't open configuration file /etc/drirc: No such file or directory.
libGL: Can't open configuration file $HOME/.drirc: No such file or directory.
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  154 (GLX)
  Minor opcode of failed request:  3 (X_GLXCreateContext)
  Value in failed request:  0x0
  Serial number of failed request:  38
  Current serial number in output stream:  40
```

I suspect it's because the nVidia driver is not installed. I tried to install it, but it complained about a driver being still loaded. I usually don't close my X session for months, as it takes me about an hour to restart everything the way I need it, so I didn't do that. Copying the nVidia libs from stretch to buster resulted in a segfault. I also tried running the resulting executable from outside the chroot, with `LD_LIBRARY_PATH` set to point to the missing libraries inside the chroot directory, but that also resulted in a segfault.

My last resort is a QEMU instance, but setting it up requires more time than a chroot. I'll try that in the next few days. My QEMU lacks OpenGL support, therefore at best I can aspire to run it with software rendering.

**Comment by Quent42340** [comment]
You also have the possibility to use a pre-installed VirtualBox/VMWare machine at https://osboxes.org, or to install a small ArchLinux instance or derivate (like Manjaro, which is really light and quick to install)

**Comment by Quent42340** [comment]
The only solution I found for Debian Stretch is to add testing repo in `/etc/apt/sources.list`:
```
deb http://deb.debian.org/debian/ testing main
```

Then

```
sudo apt update
sudo apt install gcc-7 g++-7
rm -r CMakeFiles/ CMakeCache.txt
CC=gcc-7 CXX=g++-7 cmake . && make j8
```

**Comment by pgimeno** [comment]
Every attempt to make the client run in stretch, with or without chroot, was unsuccessful. I finally made it run in a QEMU buster instance, but the mouse control was not usable due to QEMU not handling mouse capturing. There was a lot of flicker, not sure if related to the software GL renderer. As expected, rendering was pretty slow.

**Comment by pgimeno** [comment]
This is the backtrace I got from the segfault when trying the chroot-compiled version:

```bt
Thread 1 "openminer" received signal SIGSEGV, Segmentation fault.
0x000000004009dd92 in ?? ()
(gdb) bt
#0  0x000000004009dd92 in ?? ()
#1  0x00007ffff415a2d0 in ?? () from /usr/lib/libnvidia-glcore.so.340.107
#2  0x00007ffff4161360 in ?? () from /usr/lib/libnvidia-glcore.so.340.107
#3  0x00007ffff3dce378 in ?? () from /usr/lib/libnvidia-glcore.so.340.107
#4  0x00000000004e7144 in gk::RenderTarget::draw (this=0x7fffffffdd98, 
    vertexBuffer=..., mode=1, firstVertex=0, vertexCount=24, states=...)
    at /root/tmp/GameKit/source/gl/RenderTarget.cpp:32
#5  0x00000000004c6645 in BlockCursor::draw (this=0xc3bcb0, target=..., 
    states=...) at /root/tmp/OpenMiner/client/source/hud/BlockCursor.cpp:231
#6  0x00000000004e70e4 in gk::RenderTarget::draw (this=0x7fffffffdd98, 
    drawable=..., states=...)
    at /root/tmp/GameKit/source/gl/RenderTarget.cpp:26
#7  0x00000000004c8ab5 in non-virtual thunk to HUD::draw(gk::RenderTarget&, gk::RenderStates) const () at /root/tmp/OpenMiner/client/source/hud/HUD.cpp:65
#8  0x00000000004e70e4 in gk::RenderTarget::draw (this=0x7fffffffdd98, 
    drawable=..., states=...)
    at /root/tmp/GameKit/source/gl/RenderTarget.cpp:26
#9  0x00000000004cbe9e in GameState::draw (this=<optimized out>, target=..., 
    states=...) at /root/tmp/OpenMiner/client/source/states/GameState.cpp:166
#10 0x00000000004e70e4 in gk::RenderTarget::draw (this=0x7fffffffdd98, 
    drawable=..., states=...)
    at /root/tmp/GameKit/source/gl/RenderTarget.cpp:26
#11 0x00000000004dce51 in gk::CoreApplication::<lambda()>::operator()(void) const (__closure=0x7fffffffdc00)
    at /root/tmp/GameKit/source/core/CoreApplication.cpp:109
#12 0x00000000004dd260 in std::_Function_handler<void(), gk::CoreApplication::mainLoop()::<lambda()> >::_M_invoke(const std::_Any_data &) (__functor=...)
    at /usr/include/c++/6/functional:1731
#13 0x00000000004e07ec in std::function<void ()>::operator()() const (
    this=0x7fffffffdc00) at /usr/include/c++/6/functional:2127
#14 0x00000000004e0731 in gk::GameClock::drawGame(std::function<void ()>) (
    this=0x7fffffffdd58, drawFunc=...)
    at /root/tmp/GameKit/source/core/GameClock.cpp:58
#15 0x00000000004dcf37 in gk::CoreApplication::mainLoop (this=0x7fffffffdca8)
    at /root/tmp/GameKit/source/core/CoreApplication.cpp:112
#16 0x00000000004dcb3c in gk::CoreApplication::<lambda()>::operator()(void) const (__closure=0x7fffffffdc70)
    at /root/tmp/GameKit/source/core/CoreApplication.cpp:47
#17 0x00000000004dcb6c in gk::CoreApplication::run (this=0x7fffffffdca8, 
    isProtected=true) at /root/tmp/GameKit/source/core/CoreApplication.cpp:52
#18 0x00000000004ca04c in main (argc=<optimized out>, argv=<optimized out>)
    at /root/tmp/OpenMiner/client/source/main.cpp:18
```

**Comment by pgimeno** [comment]
Compilation now proceeds without problems. Running it in my system, however, causes the above segfault. I'm not sure whether the problem is with the driver or with OpenMiner, but I've managed to implement a workaround that lets me run it.

The workaround consists of not drawing the problematic stuff (highlight box and inventory cubes) during the first frame. I've implemented it in a very ugly and unreliable way, namely using a static flag that checks whether the function is called for the first time, but well, that did the trick. Anyway, the compilation problems are definitely fixed and the crash was not related to the build, therefore I'm closing this again. Thanks a lot for the help.
